rfd/0211-azure-devops-joining.md

```diff
@@ -0,0 +1,619 @@
+    // Mapped from the `rpo_ref` claim.
+    string RepositoryRef = 8 [(gogoproto.jsontag) = "repository_ref,omitempty"];
+
+  }
+  // Allow is a list of TokenRules, nodes using this token must match one
+  // allow rule to use this token.
+  repeated Rule Allow = 1 [(gogoproto.jsontag) = "allow,omitempty"];
+  // OrganizationID specifies the UUID of the Azure DevOps organization that
+  // this join token will grant access to. This is used to identify the correct
+  // issuer verification of the ID token.
+  // This is a required field.
+  string OrganizationID = 2 [(gogoproto.jsontag) = "organization_id"];
+}
+```
+
+The Join Token Create/Edit UI should be extended to support this new join
+method.
+
+#### `tbot` configuration
+
+Within `tbot` itself, no further configuration is required or will be added 
+beyond:
+
+- The user must specify the `azure_devops` join method
+- The user must specify the name of the ProvisionToken to use.
+
+However, when using `tbot` within a pipeline step, the user will need to map
+the `System.AccessToken` variable into the environment, e.g:
+
+```yaml
+steps:
+- env:
+    SYSTEM_ACCESSTOKEN: $(System.AccessToken)
+  script: tbot start -c ./my-config.yaml
+```
+
+#### Infrastructure as Code
+
+The ProvisionToken resource is already generated for the Terraform provider and
+the Kubernetes Operator, therefore the new fields to support the Azure DevOps
+join method will be included without any additional work.
+
+### Security Considerations
+
+#### Token Reuse
+
+One challenge with all OIDC join methods is the potential for token reuse.
+
+In other join methods, we leverage a challenge-response flow with a nonce within
+the `aud` claim to ensure that a token is only valid for a single join. However,
+the `aud` claim in the ID Token issued by Azure DevOps is not configurable.
+
+The Tokens do include a `jti` claim, containing a unique identifier for each
+token, which we could use to reject the reuse of a token that has already been
+used to join that particular Teleport Cluster, however, this does not mitigate
+attacks where the token was used against a third-party service
+(see Confused Deputy).
+
+To some extent, this risk is mitigated by the unusually short (5m) TTL of the 
+ID Token issued by Azure DevOps.
+
+#### `aud` Confused Deputy
+
+One challenge with the ID Tokens issued by Azure DevOps is that we cannot
+specify a custom `aud` claim for inclusion in the ID Token.
+
+This presents the risk of an ID Token intended for another service (or another
+Teleport Cluster) being reused against our Teleport Cluster if that other
+service has been compromised.
+
+To some extent, this risk is mitigated by the unusually short (5m) TTL of the 
+ID Token issued by Azure DevOps.
+
+This risk would be mitigated by the use of a Service Connection ID Token (see
+Alternatives: Use Service Connection JWTs rather than Pipeline JWTs) as users
+could be encouraged to create a Service Connection per Teleport Cluster.
```

Comment:
Out of curiosity... could it be possible to require and verify both JWTs? e.g. put the SC token in another env var, pass both along to auth, and verify both tokens? Then we'd effectively get an `aud` field, but also have full audit fields.

(not that this would be good UX by any means, it's very frustrating that this is needed at all)

Reply:
We could and I did mull this over. I came to a conclusion against it because:

- The Service Connection setup/binding to the CI is a fairly clunky bit of UX. Users would need to create and add the SC to the CI flow.
- We wouldn't have a way to guarantee both tokens came from the same run, so whilst this provides some mitigation, it's in a very /weird/ way that doesn't make much sense to me from a correctness angle. I'd rather not add extra complexity through an additional security mechanism if the mechanism isn't fully sound as this feels a bit like taping over the cracks.